packages/orm/src/Model.ts

```diff
@@ -0,0 +1,786 @@
+      throw new Error(
+        [
+          `Tried to create a ${this.modelName} model instance without a session. `,
+          "Create a session using `session = orm.session()` and call ",
+          `\`session["${this.modelName}"].create\` instead.`,
+        ].join("")
+      );
+    }
+    const props = { ...userProps };
+
+    const m2mRelations: Record<string, any> = {};
+
+    const declaredFieldNames = Object.keys(this.fields);
+    const declaredVirtualFieldNames = Object.keys(this.virtualFields);
+
+    declaredFieldNames.forEach((key) => {
+      const field = this.fields[key];
+      const valuePassed = userProps.hasOwnProperty(key);
+      if (!(field instanceof ManyToMany)) {
+        if (valuePassed) {
+          const value = userProps[key];
+          props[key] = normalizeEntity(value);
+        } else if (field.getDefault) {
+          props[key] = field.getDefault();
+        }
+      } else if (valuePassed) {
+        // If a value is supplied for a ManyToMany field,
+        // discard them from props and save for later processing.
+        m2mRelations[key] = userProps[key];
+        delete props[key];
+      }
+    });
+
+    // add backward many-many if required
+    declaredVirtualFieldNames.forEach((key) => {
+      if (!m2mRelations.hasOwnProperty(key)) {
+        const field = this.virtualFields[key];
+        if (userProps.hasOwnProperty(key) && field instanceof ManyToMany) {
+          // If a value is supplied for a ManyToMany field,
+          // discard them from props and save for later processing.
+          m2mRelations[key] = userProps[key];
+          delete props[key];
+        }
+      }
+    });
+
+    const newEntry = this.session.applyUpdate({
+      action: CREATE,
+      table: this.modelName,
+      payload: props,
+    });
+
+    const ThisModel = this;
+    const instance = new ThisModel(newEntry);
+    instance._refreshMany2Many(m2mRelations); // eslint-disable-line no-underscore-dangle
+    return instance;
+  }
+
+  /**
+   * Creates a new or update existing record in the database, instantiates a {@link Model} and returns it.
+   *
+   * If you pass values for many-to-many fields, instances are created on the through
+   * model as well.
+   *
+   * @param  {props} userProps - the required {@link Model}'s properties.
+   * @return {Model} a {@link Model} instance.
+   */
+  static upsert(userProps: { [k: string]: any }) {
+    if (typeof this.session === "undefined") {
+      throw new Error(
+        [
+          `Tried to upsert a ${this.modelName} model instance without a session. `,
+          "Create a session using `session = orm.session()` and call ",
+          `\`session["${this.modelName}"].upsert\` instead.`,
+        ].join("")
+      );
+    }
+
+    const { idAttribute } = this;
+    if (userProps.hasOwnProperty(idAttribute)) {
+      const id = userProps[idAttribute];
+      if (this.idExists(id)) {
+        const model = this.withId(id) as Model;
```

Comment:
do we need that cast here? Perhaps it'd be better to define an explicit return type for `withId` ?